pkg/controller/quota/quota.go

```diff
@@ -0,0 +1,165 @@
+package quota
+
+import (
+	"fmt"
+
+	v1 "github.com/acorn-io/acorn/pkg/apis/api.acorn.io/v1"
+	apiv1 "github.com/acorn-io/acorn/pkg/apis/internal.acorn.io/v1"
+	adminv1 "github.com/acorn-io/acorn/pkg/apis/internal.admin.acorn.io/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/acorn-io/acorn/pkg/condition"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"github.com/acorn-io/baaah/pkg/router"
+)
+
+// WaitForAllocation blocks the appInstance from being deployed until quota has been allocated on
+// an associated QuotaRequest object.
+func WaitForAllocation(req router.Request, resp router.Response) error {
+	appInstance := req.Object.(*apiv1.AppInstance)
+
+	// Create a condition setter for AppInstanceConditionQuotaAllocated, which blocks the appInstance from being deployed
+	// until quota has been allocated.
+	status := condition.Setter(appInstance, resp, apiv1.AppInstanceConditionQuotaAllocated)
+
+	// Don't do anything if quota isn't enabled for this project
+	if enforced, err := isEnforced(req, appInstance.Namespace); !enforced {
+		if err != nil {
+			status.Error(err)
+			return err
+		}
+		status.Success()
+		return nil
+	}
+
+	// Attempt to get the quotaRequest for this appInstance. It should exist with the name and namespace of the
+	// appInstance being processed.
+	quotaRequest := &adminv1.QuotaRequestInstance{}
+	err := req.Client.Get(req.Ctx, router.Key(appInstance.Namespace, appInstance.Name), quotaRequest)
+	if err != nil && !errors.IsNotFound(err) {
```

Comment:
are we igoring not found error but the quotarequest should always exist?

Reply:
QuotaRequests will not always exist, at least not immediately for AppInstances. However, in that event we enter the if tree after this line and will report the `status.Unknown()`. Doing it this way just saves some lines. What do you think though?

Reply:
Got it. I commented on the other line. The way how handler are processed will make sure that the quota object is only created after all the handlers are processed

Reply:
If i'm understanding your concern correctly, the below line will prevent the App from provisioning until it has the `QuotaAllocated` status set as to `Success`. This. will only occur if the `QuotaRequest` has been created and processed in such a way that denotes success (i.e the Spec.Resources and Status.AllocatedResources are equal).

https://github.com/acorn-io/acorn/blob/fb98a440b6e1b44e5c23636d87ed8244df162210/pkg/controller/appstatus/status.go#L41

Reply:
Ok. That's good then.